2019/02/24.md

```diff
@@ -0,0 +1,52 @@
+# Redux conventions
+
+Based on the discussion about [prefixing action
+types](https://github.com/dhis2/notes/issues/18) some conventions can be
+extracted for how to use Redux.
+
+Austin mentioned an article about [Redux best
+practices](https://medium.com/@kylpo/redux-best-practices-eef55a20cc72)
+which we can use as inspiration.
+
+## Utilities
+
+- Use [redux-actions](https://github.com/redux-utilities/redux-actions)
+  to create and handle actions
+- Use [redux-thunk](https://github.com/reduxjs/redux-thunk) to deal with
+  logic that has side-effects (async, api, etc.)
+- Use [reselect](https://github.com/reduxjs/reselect) for memoized
+  selectors
+- Use [normalizr](https://github.com/paularmstrong/normalizr) for
+  simpler reducer composition
+
+## Principles
+
+- **Reducers**: are
+  [deterministic](https://github.com/reduxjs/redux/issues/1171#issuecomment-205888533)
+  and pure
+- **Selectors**: to compute derived data
+- **Compose reducers**:
+  > Think of Redux as a tree of your reducers (like react’s render tree
+  > of components). Both are just compositions of functions.
+- All application state belongs in the Redux store.
+
+## Naming
+
+- Action name: `<NOUN>_<VERB>`
```

Comment:
reg. the verb, I think it would be good to specify that we use the **imperative** mode of the verb, i.e. `ITEM_ADD`. When I started with redux, I once read somewhere that it is better to use the past tense i.e. `ITEM_ADDED`. I think this is not common practice (anymore), but adding this line would help people like me knowing how to name their actions when working for DHIS2.

Reply:
> I once read somewhere that it is better to use the past tense i.e. `ITEM_ADDED`.

Note: I think it's not better. Action types express the intent of the user (or of another source). The intent can't be in the past tense. Also the action is dispatched before anything happens, while past tense means that it has happened already.. So I don't see the advantage of this?

Reply:
I wasn't claiming past tense was better. Only claiming that it is useful to specify which tense to use with our verbs.

Reply:
Never said you were claiming it, didn't want to imply it was your opinion, sorry if it looked like that.
Just wanted to add my opinion about this

Reply:
Added note about this.